app/controllers/application_controller.rb

```diff
@@ -5,10 +5,31 @@ class ApplicationController < ActionController::Base
   # For APIs, you may want to use :null_session instead.
   protect_from_forgery with: :exception
 
+  protected
   # Use callbacks to share common setup or constraints between actions.
   def set_vault
-    @vault = params[:vault_id] ? Vault.find_by_id(params[:vault_id]) : Vault.find_by_id(params[:id])
-    redirect_to root_url if @vault.nil? || @vault.user_username != params[:username]
+    @vault = params[:vault_id] ? Vault.find_vault(params[:vault_id]) : Vault.find_vault(params[:id])
```

Comment:
@CGA1123 `def authenticate_username ...` is called first. It checks if the user is valid and checks if the user owns the vault. Therefore, if the that pass no need to authenticate `set_vault`